### PR TITLE
New APIs for managing provider plugin caches and installing plugins into them

### PR DESCRIPTION
**Pull request by apparentlymart**
These commits introduce a new package `providercache` that is intended to encapsulate the management of filesystem directories used as local "caches" of provider plugins.

In this case by "cache" we mean a copy of a provider obtained from elsewhere that we've copied to local disk for use by Terraform commands. There are two different use-cases for this "cache" idea:

* Local cache: The `.terraform` directory in a Terraform working directory contains the main set of provider plugins that are available for use by the current Terraform configuration. This directory is the main thing we're aiming to populate as part of the provider installation process in `terraform init`, and is the only place Terraform will look to find providers for commands other than `terraform init`.
* Global cache: Users may optionally configure (in the CLI config) a _global_ cache directory. If one is configured, Terraform will treat it as a read-through cache when installing into the local cache, avoiding the need to repeatedly download the same plugin packages when several configurations all consume the same version.

I had originally intended to open a PR for this with it all integrated into `terraform init` and ready to use, but my plans were derailed by the current pandemic. To make this available as a foundation for other concurrent work, I'm instead now submitting this as _dead code_ (not used yet) with the intention of improving its test coverage and integrating it into `terraform init` in subsequent PRs.

This may be easier to review on a commit-by-commit basis, rather than all together as one big diff.

I do of course welcome any feedback as usual, but I'd also like to get this incremental step merged ASAP so we can all build around it concurrently, and so I'd prefer to save improvements that do not significantly affect the exported API of `internal/providercache` or `internal/getproviders` for incremental later PRs.

### CODE REVIEW

File: internal/providercache/dir_modify.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+	// We'll prefer to create a symlink if possible, but we'll fall back to
+	// a recursive copy if symlink creation fails. It could fail for a number
+	// of reasons, including being on Windows 8 without administrator
+	// privileges or being on a legacy filesystem like FAT that has no way
+	// to represent a symlink. (Generalized symlink support for Windows was
+	// introduced in a Windows 10 minor update.)
+	//
+	// We'd prefer to use a relative path for the symlink to reduce the risk
+	// of it being broken by moving things around later, but we'll fall back
```
Comment (alisdair) [inline internal/providercache/dir_modify.go]: I'd expect a relative path to be more likely to break than an absolute one. I'd be surprised if the configuration directory is moved along with any global cache directory—I'd expect configurations to move around, and the global cache to stay still. Do I misunderstand?

File: internal/providercache/dir_test.go [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+	randomProvider := addrs.NewProvider(
+		addrs.DefaultRegistryHost, "hashicorp", "random",
+	)
+	nonExistProvider := addrs.NewProvider(
+		addrs.DefaultRegistryHost, "bloop", "nonexist",
+	)
+	legacyProvider := addrs.NewLegacyProvider("legacy")
+
+	t.Run("ProviderLatestVersion", func(t *testing.T) {
+		t.Run("exists", func(t *testing.T) {
```
Comment (alisdair) [inline internal/providercache/dir_test.go]: I don't see test coverage for this function in the case where multiple versions of the same provider are installed. Maybe something to add later (unless it's already covered and I just missed it).

File: internal/providercache/installer_events.go [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+import (
+	"context"
+
+	"github.com/hashicorp/terraform/addrs"
+	"github.com/hashicorp/terraform/internal/getproviders"
+)
+
+// InstallerEvents is a collection of function references that can be
+// associated with an Installer object in order to be notified about various
+// installation lifecycle events during an install operation.
```
Comment (alisdair) [inline internal/providercache/installer_events.go]: This is a really nice pattern that I hadn't seen before 👏

File: internal/providercache/dir_modify.go (hunk 1 above)
Comment (apparentlymart) [reply internal/providercache/dir_modify.go]: This is a fair point... I will be honest and say that I just mimicked what our old cache codepath was doing here without thinking too critically about it. I'll switch this to be absolute in a later PR.

File: internal/providercache/dir_modify_test.go [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+	"testing"
+
+	"github.com/apparentlymart/go-versions/versions"
+	"github.com/google/go-cmp/cmp"
+
+	"github.com/hashicorp/terraform/addrs"
+	"github.com/hashicorp/terraform/internal/getproviders"
+)
+
+func LinkFromOtherCache(t *testing.T) {
```
Comment (apparentlymart) [inline internal/providercache/dir_modify_test.go]: Note to self to fix in a subsequent PR: this name doesn't have the `Test` prefix, so this test presumably isn't actually running?

File: internal/providercache/installer_events.go (hunk 3 above)
Comment (apparentlymart) [reply internal/providercache/installer_events.go]: I think the place I saw it first is [`httptrace.ClientTrace`](https://godoc.org/net/http/httptrace#ClientTrace).

**Pull request merged**